node/cn/api.go

```diff
@@ -452,3 +434,85 @@ func (api *PrivateDebugAPI) getModifiedAccounts(startBlock, endBlock *types.Bloc
+
+	startBlock = api.cn.blockchain.GetBlockByNumber(startNum)
+	if startBlock == nil {
+		return nil, nil, fmt.Errorf("start block number %d not found", startNum)
+	}
+
+	if endNum == nil {
+		endBlock = startBlock
+		startBlock = api.cn.blockchain.GetBlockByHash(startBlock.ParentHash())
+		if startBlock == nil {
+			return nil, nil, fmt.Errorf("block number %d has no parent", startNum)
+		}
+	} else {
+		endBlock = api.cn.blockchain.GetBlockByNumber(*endNum)
+		if endBlock == nil {
+			return nil, nil, fmt.Errorf("end block number %d not found", *endNum)
+		}
+	}
+
+	if startBlock.Number().Uint64() >= endBlock.Number().Uint64() {
+		return nil, nil, fmt.Errorf("start block height (%d) must be less than end block height (%d)", startBlock.Number().Uint64(), endBlock.Number().Uint64())
+	}
+
+	return startBlock, endBlock, nil
+}
+
+// GetModifiedStorageNodesByNumber returns the number of storage nodes of a contract account
+// that have been changed between the two blocks specified.
+//
+// With the first two parameters, it returns the number of storage trie nodes modified in the specified block.
+func (api *PrivateDebugAPI) GetModifiedStorageNodesByNumber(contractAddr common.Address, startNum uint64, endNum *uint64, printDetail *bool) (int, error) {
+	startBlock, endBlock, err := api.getStartAndEndBlock(startNum, endNum)
+	if err != nil {
+		return 0, err
+	}
+	return api.getModifiedStorageNodes(contractAddr, startBlock, endBlock, printDetail)
+}
+
+func (api *PrivateDebugAPI) getModifiedStorageNodes(contractAddr common.Address, startBlock, endBlock *types.Block, printDetail *bool) (int, error) {
+	startBlockRoot, err := api.cn.blockchain.GetContractStorageRoot(startBlock, api.cn.blockchain.StateCache(), contractAddr)
+	if err != nil {
+		return 0, err
+	}
+	endBlockRoot, err := api.cn.blockchain.GetContractStorageRoot(endBlock, api.cn.blockchain.StateCache(), contractAddr)
+	if err != nil {
+		return 0, err
+	}
+
+	trieDB := api.cn.blockchain.StateCache().TrieDB()
+	oldTrie, err := statedb.NewSecureTrie(startBlockRoot, trieDB)
+	if err != nil {
+		return 0, err
+	}
+	newTrie, err := statedb.NewSecureTrie(endBlockRoot, trieDB)
+	if err != nil {
+		return 0, err
+	}
+
+	diff, _ := statedb.NewDifferenceIterator(oldTrie.NodeIterator([]byte{}), newTrie.NodeIterator([]byte{}))
+	iter := statedb.NewIterator(diff)
+
+	logger.Info("Start collecting the modified storage nodes", "contractAddr", contractAddr.String(),
+		"startBlock", startBlock.NumberU64(), "endBlock", endBlock.NumberU64())
+	start := time.Now()
+	var dirty []common.Hash
+	for iter.Next() {
+		dirty = append(dirty, common.BytesToHash(iter.Key))
+	}
+	logger.Info("Finished collecting the modified storage nodes", "contractAddr", contractAddr.String(),
+		"startBlock", startBlock.NumberU64(), "endBlock", endBlock.NumberU64(), "numModifiedNodes", len(dirty), "elapsed", time.Since(start))
+	for _, d := range dirty {
+		if printDetail != nil && *printDetail {
```

Comment:
We better switch the place of these lines.

Reply:
Thanks! Updated by 542ccd6bdd43af31059060cde9a0df0db790487e